diesel/src/pg/types/network_address.rs

```diff
@@ -0,0 +1,194 @@
+extern crate ipnetwork;
+
+use std::io::prelude::*;
+use std::error::Error;
+use std::net::{Ipv4Addr, Ipv6Addr};
+
+use pg::Pg;
+use types::{self, ToSql, IsNull, FromSql};
+use self::ipnetwork::{IpNetwork,Ipv4Network,Ipv6Network};
+
+// const PGSQL_AF_INET: u8 = libc::AF_INET + 0;
+// const PGSQL_AF_INET6: u8 = libc::AF_INET + 1;
+const PGSQL_AF_INET: u8 = 2 + 0;
+const PGSQL_AF_INET6: u8 = 2 + 1;
+
+primitive_impls!(Cidr -> (IpNetwork, pg: (650, 651)));
+primitive_impls!(Inet -> (IpNetwork, pg: (869, 1041)));
+
+macro_rules! err {
+    () => (return Err("invalid network address format".into()))
+}
+
+macro_rules! assert_or_error {
+    ($cond: expr) => {
+        if ! $cond { err!() }
+    }
+}
+
+macro_rules! impl_Sql {
+    ($ty: ty, $net_type: expr) => {
+        impl FromSql<$ty, Pg> for IpNetwork {
+            fn from_sql(bytes: Option<&[u8]>) -> Result<Self, Box<Error + Send + Sync>> {
+                // https://github.com/postgres/postgres/blob/55c3391d1e6a201b5b891781d21fe682a8c64fe6/src/include/utils/inet.h#L23-L28
+                let bytes = not_none!(bytes);
+                assert_or_error!(4 <= bytes.len());
+                let af = bytes[0];
+                let prefix = bytes[1];
+                let net_type = bytes[2];
+                let len = bytes[3];
+                assert_or_error!(net_type == $net_type);
+                println!("{:?}", bytes);
+                if af == PGSQL_AF_INET {
+                    assert_or_error!(bytes.len() == 8);
+                    assert_or_error!(len == 4);
+                    let b = &bytes[4..];
+                    let addr = Ipv4Addr::new(b[0], b[1], b[2], b[3]);
+                    // FIXME: use `?` after Error is implemented for IpNetworkError
+                    let inet = Ipv4Network::new(addr, prefix).unwrap();
+                    Ok(IpNetwork::V4(inet))
+                } else if af == PGSQL_AF_INET6 {
+                    assert_or_error!(bytes.len() == 20);
+                    assert_or_error!(len == 16);
+                    let b = &bytes[4..];
+                    fn c(b1: u8, b2: u8) -> u16 {(b1 as u16) << 8 | (b2 as u16) }
+                    let addr = Ipv6Addr::new(
+                        c(b[0],  b[1]),  c(b[2],  b[3]),
+                        c(b[4],  b[5]),  c(b[6],  b[7]),
+                        c(b[8],  b[9]),  c(b[10], b[11]),
+                        c(b[12], b[13]), c(b[14], b[15]));
```

Comment:
The v4 code above I can follow, but I'm not sure what happens here. What's the `c` fn for? What format is `bytes` in? Is this the serialized [inet_struct](https://github.com/postgres/postgres/blob/55c3391d1e6a201b5b891781d21fe682a8c64fe6/src/include/utils/inet.h#L23-L28)? Given that this is an char array, there's no need to care about endianess. And `std::net::IpAddr` implements `From<[u8; 16]>` as well as `From<[u8; 4]>` (for Ipv4).